### DISCUSSION TITLE
Best practices for optionally having symfony authentication in a bundle

### DISCUSSION THREAD

**Discussion by tacman** [symfony/symfony#52327] [Q&A] [answered]
I have a menu bundle that currently requires symfony/security-bundle, since often I need to know the user's role before doing certain action (like adding an admin menu).

What is the best practice for making that optional?  

```php
use Symfony\Component\Security\Core\Authorization\AuthorizationCheckerInterface;
class X {
    public function __construct(private ?AuthorizationCheckerInterface $security = null) {}

    public function setAuthorizationChecker(AuthorizationCheckerInterface $authorizationChecker)
    {
        $this->authorizationChecker = $authorizationChecker;
    }


```

I feel like in the bundle there must be some way to say "if security is available, inject it".

In this situation, should I put security-bundle in require-dev?

Thanks.

**Comment by stof** [comment, marked as answer]
If the usage of AuthorizationCheckerInterface is optional, you should indeed not put it in `require`.

There is already a way to use `ContainerInterface::NULL_ON_INVALID_REFERENCE` when configuring arguments (XML and Yaml loaders have a syntax for that as well)

**Reply by tacman** [reply]
Thanks!!
